cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/distributed/JarCacheTracker.java

```diff
@@ -0,0 +1,179 @@
+   */
+  private class CacheInfo {
+    private final AtomicLong firstLaunchTime = new AtomicLong(-1L);
+
+    // synchronized here instead of at registerLaunch so that we only per program type.
+    private synchronized void waitForCacheFiles(File cacheDir) {
+      // check this in case we called this method before the files were available, were blocked on another thread,
+      // and are now here
+      long threshold = firstLaunchTime.get() + TimeUnit.MILLISECONDS.convert(90, TimeUnit.SECONDS);
+      if (System.currentTimeMillis() > threshold || cacheFilesExist(cacheDir)) {
+        return;
+      }
+
+      // otherwise, wait for at most threshold seconds for the cache files to exist
+      WatchService watcher;
+      WatchKey key;
+      try {
+        watcher = FileSystems.getDefault().newWatchService();
+        key = cacheDir.toPath().register(watcher, StandardWatchEventKinds.ENTRY_CREATE);
+      } catch (Exception e) {
+        LOG.error("Error watching {} for twill cache files to exist. Continuing on.", cacheDir, e);
+        return;
+      }
+
+      // if the cache files were written while we were registering the watcher
+      Path containerDonePath = FileSystems.getDefault().getPath(cacheDir.getAbsolutePath(), "container.jar.done");
+      Path appmasterDonePath = FileSystems.getDefault().getPath(cacheDir.getAbsolutePath(), "appMaster.jar.done");
+      boolean containerJarExists = Files.exists(containerDonePath);
+      boolean appmasterJarExists = Files.exists(appmasterDonePath);
+      if (containerJarExists && appmasterJarExists) {
+        LOG.debug("files exist, cancelling watch key.");
+        key.cancel();
+        return;
+      }
+
+      LOG.debug("Waiting for twill cache files at {}", cacheDir);
+      while (!containerJarExists || !appmasterJarExists) {
+
+        // wait at most threshold millis for some event to occur
+        long maxWaitTime = threshold - System.currentTimeMillis();
+        if (maxWaitTime < 0) {
+          break;
+        }
+        try {
+          key = watcher.poll(maxWaitTime, TimeUnit.MILLISECONDS);
+        } catch (Exception e) {
+          LOG.warn("Error while watching for cache jars. Continuing on.", e);
+          return;
+        }
+        if (key == null) {
+          LOG.debug("Waited past the threshold and cache jars are still not there. Continuing on.");
+          return;
+        }
+
+        // examine all events on the directory
+        for (WatchEvent<?> event : key.pollEvents()) {
+          WatchEvent.Kind<?> kind = event.kind();
+
+          // apparently overflow is possible regardless of what events were registered
+          if (kind == StandardWatchEventKinds.OVERFLOW) {
+            continue;
+          }
+
+          //noinspection unchecked
+          WatchEvent<Path> ev = (WatchEvent<Path>) event;
+          String filename = ev.context().getFileName().toString();
```

Comment:
Why not compare to the `Path`, `containerDonePath` and `appmasterDonePath`? No need to have toString.